cdc/generation.cc

```diff
@@ -173,118 +174,17 @@ topology_description generate_topology_description(
-        // We were not able to generate stream ids for each (token range, shard) pair.
-
-        // For each range that has a stream, for each shard for this range that doesn't have a stream,
-        // use the stream id of the next shard for this range.
-
-        // For each range that doesn't have any stream,
-        // use streams of the first range to the left which does have a stream.
-
-        cdc_log.warn("Generation of CDC streams failed to create streams for some (vnode, shard) pair."
-                     " This can lead to worse performance.");
-
-        stream_id some_stream;
-        size_t idx = 0;
-        for (; idx < entries.size(); ++idx) {
-            for (auto s: entries[idx].streams) {
-                if (s.is_set()) {
-                    some_stream = s;
-                    break;
-                }
-            }
-            if (some_stream.is_set()) {
-                break;
-            }
-        }
-
-        assert(idx != entries.size() && some_stream.is_set());
-
-        // Iterate over all ranges in the clockwise direction, starting with the one we found a stream for.
-        for (size_t off = 0; off < entries.size(); ++off) {
-            auto& ss = entries[(idx + off) % entries.size()].streams;
-
-            int last_set_stream_idx = ss.size() - 1;
-            while (last_set_stream_idx > -1 && !ss[last_set_stream_idx].is_set()) {
-                --last_set_stream_idx;
-            }
-
-            if (last_set_stream_idx == -1) {
-                cdc_log.warn(
-                        "CDC wasn't able to generate any stream for vnode ({}, {}]. We'll use another vnode's streams"
-                        " instead. This might lead to inconsistencies.",
-                        tokens[(idx + off + entries.size() - 1) % entries.size()], tokens[(idx + off) % entries.size()]);
-
-                ss[0] = some_stream;
-                last_set_stream_idx = 0;
-            }
-
-            some_stream = ss[last_set_stream_idx];
-
-            // Replace 'unset' stream ids with indexes below last_set_stream_idx
-            for (int s_idx = last_set_stream_idx - 1; s_idx > -1; --s_idx) {
-                if (ss[s_idx].is_set()) {
-                    some_stream = ss[s_idx];
-                } else {
-                    ss[s_idx] = some_stream;
-                }
-            }
-            // Replace 'unset' stream ids with indexes above last_set_stream_idx
-            for (int s_idx = ss.size() - 1; s_idx > last_set_stream_idx; --s_idx) {
-                if (ss[s_idx].is_set()) {
-                    some_stream = ss[s_idx];
-                } else {
-                    ss[s_idx] = some_stream;
-                }
+        dht::sharder sharder(entry.streams.size(), entry.sharding_ignore_msb);
+        for (size_t shard_idx = 0; shard_idx < entry.streams.size(); ++shard_idx) {
+            auto t = sharder.token_for_next_shard(prev_token, shard_idx);
+            if (t > entry.token_range_end) {
```

Comment:
In addition, this test can fail even though we got a good token. Consider min_token = 0 and max_token = 1000. Consider prev_token = 990 and t = 10 (a wraparound range). Suppose token_for_next_shard returns 995. Then this test will compare it against 10 and fail, even though 995 is inside the wraparound range (990, 10].

Reply:
I think you can fix both problems by using ring_position_range_vector_sharder. Use a vector of ranges instead of a range; the wraparound range translates to a two-element vector and the rest translate to a single element vector. You can keep calling ring_position_range_vector_sharder_oh_why_is_this_so_long::next() until you have enough for smp::count or until it return nullopt, indicating no more shards own data.

Reply:
That would require some unnecessary allocations. I solved both issues by just special casing vnode (tokens.back(), tokens.front()].

Reply:
This happens once per bootstrap, allocations here are not a problem, but correctness is. I hope this isn't causing more complications.

Reply:
It's actually quite simple. I even think that the resulting version is easier to understand.